dpnp/backend/extensions/indexing/choose.cpp

```diff
@@ -0,0 +1,444 @@
+    return param_pack_deps;
+}
+
+// copied from dpctl, remove if a similar utility is ever exposed
+std::vector<dpctl::tensor::usm_ndarray> parse_py_chcs(const sycl::queue &q,
+                                                      const py::object &py_chcs)
+{
+    py::ssize_t chc_count = py::len(py_chcs);
+    std::vector<dpctl::tensor::usm_ndarray> res;
+    res.reserve(chc_count);
+
+    for (py::ssize_t i = 0; i < chc_count; ++i) {
+        py::object el_i = py_chcs[py::cast(i)];
+        dpctl::tensor::usm_ndarray arr_i =
+            py::cast<dpctl::tensor::usm_ndarray>(el_i);
+        if (!dpctl::utils::queues_are_compatible(q, {arr_i})) {
+            throw py::value_error("Choice allocation queue is not compatible "
+                                  "with execution queue");
+        }
+        res.push_back(arr_i);
+    }
+
+    return res;
+}
+
+std::pair<sycl::event, sycl::event>
+    py_choose(const dpctl::tensor::usm_ndarray &src,
+              const py::object &py_chcs,
+              const dpctl::tensor::usm_ndarray &dst,
+              uint8_t mode,
+              sycl::queue &exec_q,
+              const std::vector<sycl::event> &depends)
+{
+    std::vector<dpctl::tensor::usm_ndarray> chcs =
+        parse_py_chcs(exec_q, py_chcs);
+
+    // Python list max size must fit into py_ssize_t
+    py::ssize_t n_chcs = chcs.size();
+
+    if (n_chcs == 0) {
+        throw py::value_error("List of choices is empty.");
+    }
+
+    if (mode != 0 && mode != 1) {
+        throw py::value_error("Mode must be 0 or 1.");
+    }
+
+    dpctl::tensor::validation::CheckWritable::throw_if_not_writable(dst);
+
+    const dpctl::tensor::usm_ndarray chc_rep = chcs[0];
+
+    int nd = src.get_ndim();
+    int dst_nd = dst.get_ndim();
+    int chc_nd = chc_rep.get_ndim();
+
+    if (nd != dst_nd || nd != chc_nd) {
+        throw py::value_error("Array shapes are not consistent");
+    }
+
+    const py::ssize_t *src_shape = src.get_shape_raw();
+    const py::ssize_t *dst_shape = dst.get_shape_raw();
+    const py::ssize_t *chc_shape = chc_rep.get_shape_raw();
+
+    size_t nelems = src.get_size();
+    bool shapes_equal = std::equal(src_shape, src_shape + nd, dst_shape);
+    shapes_equal &= std::equal(src_shape, src_shape + nd, chc_shape);
+
+    if (!shapes_equal) {
+        throw py::value_error("Array shapes don't match.");
+    }
+
+    if (nelems == 0) {
+        return std::make_pair(sycl::event{}, sycl::event{});
+    }
+
+    char *src_data = src.get_data();
+    char *dst_data = dst.get_data();
+
+    if (!dpctl::utils::queues_are_compatible(exec_q, {src, dst})) {
+        throw py::value_error(
+            "Execution queue is not compatible with allocation queues");
+    }
+
+    auto const &overlap = dpctl::tensor::overlap::MemoryOverlap();
+    if (overlap(src, dst)) {
+        throw py::value_error("Array memory overlap.");
+    }
+
+    py::ssize_t src_offset = py::ssize_t(0);
+    py::ssize_t dst_offset = py::ssize_t(0);
```

Comment:
Seems they are set to zero and will never be modified. Is it expected?

Reply:
If preferred, I can remove `src_offset` and `dst_offset` and instead use `py::ssize_t(0)` in the impl call.